cpp/include/cudf/io/orc.hpp

```diff
@@ -0,0 +1,832 @@
+   *
+   * @param val Boolean value to enable/disable.
+   * @return this for chaining.
+   */
+  chunked_orc_writer_options_builder& enable_statistics(bool val)
+  {
+    options._enable_statistics = val;
+    return *this;
+  }
+
+  /**
+   * @brief Sets associated metadata.
+   *
+   * @param meta Associated metadata.
+   * @return this for chaining.
+   */
+  chunked_orc_writer_options_builder& metadata(table_metadata_with_nullability* meta)
+  {
+    options._metadata = meta;
+    return *this;
+  }
+
+  /**
+   * @brief move chunked_orc_writer_options member once it's built.
+   */
+  operator chunked_orc_writer_options &&() { return std::move(options); }
+
+  /**
+   * @brief move chunked_orc_writer_options member once it's built.
+   *
+   * This has been added since Cython does not support overloading of conversion operators.
+   */
+  chunked_orc_writer_options&& build() { return std::move(options); }
+};
+
+/**
+ * @brief Forward declaration of anonymous chunked-writer state struct.
+ */
+struct orc_chunked_state;
+
+/**
+ * @brief Begin the process of writing an ORC file in a chunked/stream form.
+ *
+ * @ingroup io_writers
+ *
+ * The intent of the write_orc_chunked_ path is to allow writing of an
+ * arbitrarily large / arbitrary number of rows to an ORC file in multiple passes.
+ *
+ * The following code snippet demonstrates how to write a single ORC file containing
+ * one logical table by writing a series of individual cudf::tables.
+ * @code
+ *  ...
+ *  std::string filepath = "dataset.orc";
+ *  cudf::io::chunked_orc_writer_options options = cudf::io::chunked_orc_writer_options
+ * options::builder(cudf::sink_info(filepath));
+ *  ...
+ *  auto state = cudf::write_orc_chunked_begin(options);
+ *    cudf::write_orc_chunked(table0, state);
+ *    cudf::write_orc_chunked(table1, state);
+ *    ...
+ *  cudf_write_orc_chunked_end(state);
+ * @endcode
+ *
+ * @param[in] options Settings for controlling writing behavior.
+ * @param[in] mr Device memory resource to use for device memory allocation.
+ *
+ * @returns pointer to an anonymous state structure storing information about the chunked write.
+ * this pointer must be passed to all subsequent write_orc_chunked() and write_orc_chunked_end()
+ * calls.
+ */
+std::shared_ptr<orc_chunked_state> write_orc_chunked_begin(
+  chunked_orc_writer_options const& options,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Write a single table as a subtable of a larger logical orc file/table.
+ *
+ * @ingroup io_writers
+ *
+ * All tables passed into multiple calls of this function must contain the same # of columns and
+ * have columns of the same type.
+ *
+ * @param[in] table The table data to be written.
+ * @param[in] state Opaque state information about the writer process. Must be the same pointer
+ * returned from write_orc_chunked_begin().
+ */
+void write_orc_chunked(table_view const& table, std::shared_ptr<orc_chunked_state> state);
```

Comment:
should we make `state` be the first parameter, so it's consistent with `write_orc_chunked_end`?

Reply:
Shouldn't be a problem